mbed_targets/_internal/target_database.py

```diff
@@ -2,41 +2,65 @@
 
 import logging
 from json.decoder import JSONDecodeError
-from typing import List, Optional
+from typing import List
 
 import requests
 
 logger = logging.getLogger(__name__)
 
 _AUTH_TOKEN_ENV_VAR = "MBED_API_AUTH_TOKEN"
-
 _TARGET_API = "https://os.mbed.com/api/v4/targets"
 
 
-def get_target_data() -> Optional[List[dict]]:
+class TargetAPIError(Exception):
+    """API request failed."""
+
+
+class ResponseJSONError(Exception):
+    """HTTP reponse JSON parsing failed."""
+
+
+def get_target_data() -> List[dict]:
     """Retrieves list of build targets and OS versions to determine versions of the OS to test against.
 
     Returns:
-        The target database as retrieved from the targets API, None if there was an error.
-    """
-    target_data = None
+        The target database as retrieved from the targets API
 
+    Raises:
+        ResponseJSONError: error decoding the reponse JSON.
+        TargetAPIError: error retrieving data from the Target API.
+    """
+    target_data: List[dict] = [{}]
     response = requests.get(_TARGET_API)
+    _check_response_status_code(response)
```

Comment:
I can understand the change in flow in performs the error handling in one place, but the multiple nested function calls have obfuscated what was a clear flow of control.

Reply:
I disagree; the previous flow of control was quite hard to follow with the nested `try/except` and `if/else` statements; the congnitive complexity of the function in its previous form was too high for my liking (and code climate seemed to agree). I think this change makes the control flow easier to understand, generally flat is better than nested. Also using `check_` in the function name usually implies that an exception will be thrown if the function fails. I can add a comment to the `_check_response_status_code` to make things a bit clearer, though.

Reply:
Lets chat about this.